### PR TITLE
Use metabase-lib v2 to manage limit in the notebook editor

### PR DESCRIPTION
**Pull request by kulyk**
Epic #28034

Makes the notebook editor's `LimitStep` use metabase-lib v2 under the hood. Also, lays the foundation for porting other notebook steps to v2

### To Verify

1. Open `/question/:id/notebook`
2. Click the "Row limit" button to open the input
3. Try setting different row limit values, try different kind of queries, try multistage queries
4. Ensure everything works as on `master`

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/29618)

### CODE REVIEW

File: frontend/src/metabase-lib/limit.ts [hunk 1]
```diff
@@ -1,15 +1,21 @@
...
+): Limit {
+  return ML.current_limit(query, stageIndex);
 }
 
-export function hasLimit(query: Query) {
-  const limit = currentLimit(query);
+declare function LimitFn(query: Query, limit: Limit): Query;
+declare function LimitFn(query: Query, stageIndex: number, limit: Limit): Query;
+
+export const limit: typeof LimitFn = ML.limit;
```
Comment (kulyk) [inline frontend/src/metabase-lib/limit.ts:16]: This allows using `limit` with or without a query stage index provided:

```ts
Lib.limit(query, 5) // Set query limit to 5
Lib.limit(query, 2, 5) // Set limit to 5 for a query at stage 2
```

Using an optional parameter is not great here for two reasons:

1. We can't have an optional argument between two required args, and a signature `query, limit, stage` looks odd
2. When passed to Clojure, `undefined` is going to be treated like `nil`, and MLv2 doesn't handle it well yet

File: frontend/src/metabase/query_builder/components/notebook/lib/steps.ts [hunk 2]
```diff
@@ -94,8 +96,15 @@ const STEPS: NotebookStepDef[] = [
...
-    revert: query => query.clearLimit(),
+    active: (legacyQuery, itemIndex, query, stageIndex) =>
+      Lib.hasLimit(query, stageIndex),
+    revert: (legacyQuery, itemIndex, query, stageIndex) => {
+      const reverted = Lib.limit(query, stageIndex, null);
+      const revertedLegacyQuery = Lib.toLegacyQuery(reverted);
+      const revertedStructuredQuery =
+        legacyQuery.setDatasetQuery(revertedLegacyQuery);
+      return revertedStructuredQuery.queries()[stageIndex];
+    },
```
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/notebook/lib/steps.ts:107]: I don't love making notebook step functions receive four args now, but that felt like the most sane way to keep both v1 and v2 patterns in place

**Review by bshepherdson — APPROVED** [review]
> Reviewed 13 of 13 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29618)* status: all files reviewed, 2 unresolved discussions (waiting on @alxnddr, @iethree, and @ranquild)

**Review by ranquild — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29618)* status: all files reviewed, 3 unresolved discussions (waiting on @alxnddr and @kulyk)

___
*[`frontend/src/metabase/query_builder/components/notebook/steps/LimitStep/LimitStep.unit.spec.tsx` line 72 at r1](https://reviewable.io/reviews/metabase/metabase/29618#-NRnMi6h-S1_5ciIDT2t:-NRnMi6h-S1_5ciIDT2u:b4k0k5s) ([raw file](https://github.com/metabase/metabase/blob/0d00dc01031a7271375eb7a9b4eb9117f0c3744e/frontend/src/metabase/query_builder/components/notebook/steps/LimitStep/LimitStep.unit.spec.tsx#L72)):*
> ```XML
> 
>     const limitInput = screen.getByPlaceholderText("Enter a limit");
>     fireEvent.change(limitInput, { target: { value: "0" } });
> ```

Why not userEvent?

**Comment by kulyk** [comment]
*[`frontend/src/metabase/query_builder/components/notebook/steps/LimitStep/LimitStep.unit.spec.tsx` line 72 at r1](https://reviewable.io/reviews/metabase/metabase/29618#-NRnMi6h-S1_5ciIDT2t:-NRnN5kv0ITbhapcIlHk:b-9t0dgh) ([raw file](https://github.com/metabase/metabase/blob/0d00dc01031a7271375eb7a9b4eb9117f0c3744e/frontend/src/metabase/query_builder/components/notebook/steps/LimitStep/LimitStep.unit.spec.tsx#L72)):*
<details><summary><i>Previously, ranquild (Alexander Polyankin) wrote…</i></summary><blockquote>

Why not userEvent?
</blockquote></details>

@ranquild, I'm not sure why, but it works weirdly for this input. For example, when doing something like:

```js
userEvent.type(limitInput, "52")
```

it will fire two `onChange` events with "5" and "2" values. In the real world, `onChange` is called once with "52" :/

**Pull request merged**